### ISSUE TITLE
Stopping a container brings down the host eth0 interface

### ISSUE THREAD

**Issue by rgl** [lxc/incus#146]
# Required information

 * Distribution: Debian
 * Distribution version: 12
 * The output of "inc info" or if that fails:
   * Kernel version: 6.x
   * LXC version: n/a
   * Incus version: 0.1
   * Storage backend in use: btrfs

# Issue description

As discussed in https://github.com/zabbly/incus/issues/5, stopping a container brings down the host `eth0` interface.

Renaming the host `eth0` to the default `enp5s0` works (actually, I've removed `net.ifnames=0` from the linux kernel command line); when stopping the container, it no longer shuts down the host `enp5s0` network interface.

Please note that this does not happen when using a virtual machine; in this case, the host interface is not shutdown.

# Steps to reproduce

With Incus running in a Debian 12 (vagrant managed) virtual machine, and configured as https://github.com/rgl/incus-playground/blob/main/provision-incus.sh#L71-L99, follow the next steps.

```console
# ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:54:00:8f:68:65 brd ff:ff:ff:ff:ff:ff
    altname enp5s0
    inet 192.168.121.125/24 brd 192.168.121.255 scope global dynamic eth0
       valid_lft 3384sec preferred_lft 3384sec
    inet6 fe80::5054:ff:fe8f:6865/64 scope link 
       valid_lft forever preferred_lft forever
3: incusbr0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether 00:16:3e:b3:a2:02 brd ff:ff:ff:ff:ff:ff
    inet 10.2.0.1/24 scope global incusbr0
       valid_lft forever preferred_lft forever

# incus launch images:debian/12 debian-ct
Creating debian-ct
Starting debian-ct                            

# ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 52:54:00:8f:68:65 brd ff:ff:ff:ff:ff:ff
    altname enp5s0
    inet 192.168.121.125/24 brd 192.168.121.255 scope global dynamic eth0
       valid_lft 3357sec preferred_lft 3357sec
    inet6 fe80::5054:ff:fe8f:6865/64 scope link 
       valid_lft forever preferred_lft forever
3: incusbr0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 00:16:3e:b3:a2:02 brd ff:ff:ff:ff:ff:ff
    inet 10.2.0.1/24 scope global incusbr0
       valid_lft forever preferred_lft forever
5: vethc085fdca@if4: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master incusbr0 state UP group default qlen 1000
    link/ether c6:48:85:2f:67:d5 brd ff:ff:ff:ff:ff:ff link-netnsid 0

# incus stop debian-ct

# ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST> mtu 1500 qdisc fq_codel state DOWN group default qlen 1000
    link/ether 52:54:00:8f:68:65 brd ff:ff:ff:ff:ff:ff
    altname enp5s0
3: incusbr0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether 00:16:3e:b3:a2:02 brd ff:ff:ff:ff:ff:ff
    inet 10.2.0.1/24 scope global incusbr0
       valid_lft forever preferred_lft forever

# journalctl
...
Oct 08 09:14:05 incus kernel: incusbr0: port 1(vethc085fdca) entered disabled state
Oct 08 09:14:05 incus kernel: veth0ed2f97b: renamed from physpu5wB2
Oct 08 09:14:05 incus (udev-worker)[13609]: Network interface NamePolicy= disabled on kernel command line.
Oct 08 09:14:05 incus kernel: device vethc085fdca left promiscuous mode
Oct 08 09:14:05 incus kernel: incusbr0: port 1(vethc085fdca) entered disabled state
Oct 08 09:14:05 incus systemd[1]: Stopping ifup@eth0.service - ifup for eth0...
Oct 08 09:14:05 incus dhclient[13642]: Killed old client process
Oct 08 09:14:05 incus ifdown[13642]: Killed old client process
Oct 08 09:14:05 incus audit[13652]: AVC apparmor="STATUS" operation="profile_remove" profile="unconfined" name="incus-debian-ct_</var/lib/incus>" pid=13652 comm="apparmor_parser"
Oct 08 09:14:05 incus kernel: audit: type=1400 audit(1696752845.948:19): apparmor="STATUS" operation="profile_remove" profile="unconfined" name="incus-debian-ct_</var/lib/incus>" pid=13652 comm="apparmor_parser"
Oct 08 09:14:06 incus dhclient[13642]: Internet Systems Consortium DHCP Client 4.4.3-P1
Oct 08 09:14:06 incus ifdown[13642]: Internet Systems Consortium DHCP Client 4.4.3-P1
Oct 08 09:14:06 incus ifdown[13642]: Copyright 2004-2022 Internet Systems Consortium.
Oct 08 09:14:06 incus ifdown[13642]: All rights reserved.
Oct 08 09:14:06 incus ifdown[13642]: For info, please visit https://www.isc.org/software/dhcp/
Oct 08 09:14:06 incus dhclient[13642]: Copyright 2004-2022 Internet Systems Consortium.
Oct 08 09:14:06 incus dhclient[13642]: All rights reserved.
Oct 08 09:14:06 incus dhclient[13642]: For info, please visit https://www.isc.org/software/dhcp/
Oct 08 09:14:06 incus dhclient[13642]: 
Oct 08 09:14:06 incus dhclient[13642]: Listening on LPF/eth0/52:54:00:8f:68:65
Oct 08 09:14:06 incus ifdown[13642]: Listening on LPF/eth0/52:54:00:8f:68:65
Oct 08 09:14:06 incus ifdown[13642]: Sending on   LPF/eth0/52:54:00:8f:68:65
Oct 08 09:14:06 incus ifdown[13642]: Sending on   Socket/fallback
Oct 08 09:14:06 incus dhclient[13642]: Sending on   LPF/eth0/52:54:00:8f:68:65
Oct 08 09:14:06 incus dhclient[13642]: Sending on   Socket/fallback
Oct 08 09:14:06 incus dhclient[13642]: DHCPRELEASE of 192.168.121.125 on eth0 to 192.168.121.1 port 67
Oct 08 09:14:06 incus ifdown[13642]: DHCPRELEASE of 192.168.121.125 on eth0 to 192.168.121.1 port 67
Oct 08 09:14:06 incus systemd[1]: ifup@eth0.service: Deactivated successfully.
Oct 08 09:14:06 incus systemd[1]: Stopped ifup@eth0.service - ifup for eth0.
```

**Comment by stgraber** [comment]
I can't seem to reproduce the issue here, which makes me suspect that it may not be a problem with Incus so much as your host system having some kind of udev rule that's misfiring.

**Comment by stgraber** [comment]
I've been testing with Debian 12 on the host (also a VM) with the network card named `eth0` and I'm not seeing any problem, but the difference likely is that I'm using systemd-networkd and not ifupdown.

**Comment by stgraber** [comment]
Yeah, confirmed, the `ifupdown` package ships with:
```
root@v1:~# cat /lib/udev/rules.d/80-ifupdown.rules
# Allow rfkill for users in the netdev group
KERNEL=="rfkill", MODE="0664", GROUP="netdev"

# Handle allow-hotplug interfaces
SUBSYSTEM=="net", ACTION=="add|remove", RUN+="ifupdown-hotplug"
```

The `ifupdown-hotplug` script then calls `ifdown ${NAME}`.

At no point is there any check for the ifindex or namespace to distinguish a kernel event for a container vs one for a host device.

So this is an ifupdown bug unfortunately.

**Comment by rgl** [comment]
Thank you for looking into it!

I've replaced ifupdown with systemd-networkd and it seems to work.

Can you please look at https://github.com/rgl/incus-playground/commit/de03a2a8f48d9d63999e6fc786d9688ba18dc5a8 and see if I'm missing anything else?

**Comment by DanielG** [comment]
I'd like to report/fix this in Debian ifupdown. Do you have any pointers on how to reproduce this without a full incus deployment?

Thanks,
--Daniel

**Comment by stgraber** [comment]
Looking at getting a simpler reproducer of what's going on.

**Comment by DanielG** [comment]
Two more loose ends: For ifupdown-hotplug to be at fault the interface must have been declared `allow-hotplug` (not just `auto`) and there's a check in net_ifdown that will fail if systemd is in-use (which seems unlikely if you were testing on bookworm.

```
net_ifdown() {
    check_program ifdown

    # systemd will automatically ifdown the interface on device
    # removal by binding the instanced service to the network device
    if [ -d /run/systemd/system ]; then
        exit 0
    fi

    exec ifdown --allow=hotplug $INTERFACE
}
```

**Comment by stgraber** [comment]
Yeah, so far I'm unable to reproduce the failure described by @rgl on bookworm with ifupdown.

I do see an event that could cause confusion for ifupdown's udev rules though, basically during container teardown the `eth0` interface of the container is relocated and renamed on the host.

This causes this:
```
KERNEL[233.296285] add      /devices/virtual/net/eth0 (net)
KERNEL[233.296313] move     /devices/virtual/net/physzpiiWQ (net)
KERNEL[233.296739] move     /devices/virtual/net/veth599cc18c (net)
```

My setup didn't use allow-hotplug and does use systemd so nothing bad happens, but let me try with allow-hotplug see if bad things happen then.

**Comment by stgraber** [comment]
Yeah, still no luck even with that.

@rgl can you please show:
 - uname -a
 - systemctl -a | grep eth0
 - cat /etc/network/interfaces
 - cat /etc/network/interfaces.d/*

And also it'd be great if you could run `udevadm monitor` when stopping a container.
All that on a system using eth0 as the host device please.

**Comment by DanielG** [comment]
Ah, I see. So if I understand incus is doing the equivalent of:

```
$ ip -netns some-netns link set dev eth0 netns 1 # (move to netns of PID 1 on host)
$ ip link set dev eth0 name whatever123 
```

Seems the renaming should take place in the netns or conflicts are bound to happen.

**Comment by stgraber** [comment]
It's pretty deep in liblxc from what I recall, but the rename is effectively all handled through a single netlink operation so the conflict never actually happens, which is why you're seeing an `add` event even though `eth0` already exists.

Not sure if @brauner remembers the details on that one. The sequence of uevents as shown above could likely be considered a kernel bug given that in practice `eth0` was never really added...

**Comment by DanielG** [comment]
A quick grep over the code reveals lxc_network_move_created_netdev_priv and the associate doc comment https://github.com/lxc/lxc/blob/main/src/lxc/network.c#L3434

From the logging calls it seems lxc_restore_phys_nics_to_netns is the code we should be looking at. https://github.com/lxc/lxc/blob/main/src/lxc/network.c#L3736

**Comment by DanielG** [comment]
So the move is indeed done in one netlink call with lxc_netdev_move_by_index_fd. The fd argument is the netns. lxc_restore_phys_nics_to_netns calls this as follows, so in principle it seems to be as you said:

```
		/* Restore physical interfaces to host's network namespace with its transient name
		 * to avoid collisions with the host's other interfaces.
		 */
		ret = lxc_netdev_move_by_index_fd(netdev->ifindex, oldfd, netdev->transient_name);
```

**Comment by DanielG** [comment]
Here's an iproute2 reproducer for a kernel bug report:

```
# ip netns add test
# ip link add dev dummy0 netns test type dummy
# ip -netns test link set dev dummy0 netns 1 name dummy123
```

udevadm monitor prints this after the last command

```
KERNEL[230953.424834] add      /devices/virtual/net/dummy0 (net)
KERNEL[230953.424914] move     /devices/virtual/net/dummy0 (net)
KERNEL[230953.425066] move     /devices/virtual/net/dummy123 (net)
UDEV  [230953.481852] add      /devices/virtual/net/dummy0 (net)
UDEV  [230953.499799] move     /devices/virtual/net/dummy0 (net)
UDEV  [230953.500915] move     /devices/virtual/net/dummy123 (net)
```

--Daniel

**Comment by stgraber** [comment]
Right, so that indeed matches the behavior we're seeing, now the question is why would this cause the network to go down. I can see how it may re-trigger `ifup` but that should just immediately fail due to it already being up, it shouldn't cause it to go down.

**Comment by DanielG** [comment]
Having another look at the initial report's log I see that the systemd unit ifup@eth0 is what is calling ifdown. So you had the cause right, just not the mechanism. ifupdown-hotplug is only relevant on sysvinit based systems but when the ifup@.service unit is in play it has the same effect through the `BindsTo=sys-subsystem-net-devices-%i.device` stanza.

My guess is systemd treats the eth0 -> physXXX rename event as stop for the device unit. see systemd.device(5), but this needs more testing to be sure.

**Comment by DanielG** [comment]
We have a winner (probably):

```
static int device_dispatch_io(sd_device_monitor *monitor, sd_device *dev, void *userdata) {
[...]
        if (action == SD_DEVICE_MOVE)
                device_remove_old_on_move(m, dev);
```

**Comment by stgraber** [comment]
Ah, that'd explain it indeed, so yeah, basically feels like a kernel bug.
Feels like we should either get a single `add` with the new host-side name or a `move` with the same effect.

**Comment by rgl** [comment]
Thanks for looking into this! This was quite an interesting read.

Here's the requested information.

```console
# uname -a
Linux incus 6.1.0-13-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.55-1 (2023-09-29) x86_64 GNU/Linux

# systemctl -a | grep eth0
sys-devices-pci0000:00-0000:00:02.4-0000:05:00.0-virtio3-net-eth0.device  loaded    active   plugged   Virtio 1.0 network device
sys-subsystem-net-devices-eth0.device                                     loaded    active   plugged   Virtio 1.0 network device
ifup@eth0.service                                                         loaded    active   exited    ifup for eth0
```

## /etc/network/interfaces

```bash
# This file describes the network interfaces available on your system
# and how to activate them. For more information, see interfaces(5).

source /etc/network/interfaces.d/*

# The loopback network interface
auto lo
iface lo inet loopback

# The primary network interface
allow-hotplug eth0
iface eth0 inet dhcp
auto eth1
iface eth1 inet manual
auto eth2
iface eth2 inet manual
auto eth3
iface eth3 inet manual
```

## /etc/network/interfaces.d/*

there are no files here.

## udevadm monitor

Captured while `incus launch images:debian/12 debian-ct` is running:

```console
# udevadm monitor
monitor will print the received events for:
UDEV - the event which udev sends out after rule processing
KERNEL - the kernel uevent

KERNEL[1113.710951] add      /module/veth (module)
KERNEL[1113.712619] add      /devices/virtual/net/vethe5e8bca1 (net)
KERNEL[1113.713133] add      /devices/virtual/net/vethe5e8bca1/queues/rx-0 (queues)
KERNEL[1113.713504] add      /devices/virtual/net/vethe5e8bca1/queues/rx-1 (queues)
KERNEL[1113.714143] add      /devices/virtual/net/vethe5e8bca1/queues/rx-2 (queues)
KERNEL[1113.714506] add      /devices/virtual/net/vethe5e8bca1/queues/rx-3 (queues)
KERNEL[1113.714867] add      /devices/virtual/net/vethe5e8bca1/queues/tx-0 (queues)
KERNEL[1113.715259] add      /devices/virtual/net/vethe5e8bca1/queues/tx-1 (queues)
KERNEL[1113.715627] add      /devices/virtual/net/vethe5e8bca1/queues/tx-2 (queues)
KERNEL[1113.715994] add      /devices/virtual/net/vethe5e8bca1/queues/tx-3 (queues)
KERNEL[1113.716961] add      /devices/virtual/net/vethee346c56 (net)
KERNEL[1113.717312] add      /devices/virtual/net/vethee346c56/queues/rx-0 (queues)
KERNEL[1113.719270] add      /devices/virtual/net/vethee346c56/queues/rx-1 (queues)
KERNEL[1113.719512] add      /devices/virtual/net/vethee346c56/queues/rx-2 (queues)
KERNEL[1113.719618] add      /devices/virtual/net/vethee346c56/queues/rx-3 (queues)
KERNEL[1113.719945] add      /devices/virtual/net/vethee346c56/queues/tx-0 (queues)
KERNEL[1113.720137] add      /devices/virtual/net/vethee346c56/queues/tx-1 (queues)
KERNEL[1113.720481] add      /devices/virtual/net/vethee346c56/queues/tx-2 (queues)
KERNEL[1113.720615] add      /devices/virtual/net/vethee346c56/queues/tx-3 (queues)
UDEV  [1113.720866] add      /module/veth (module)
KERNEL[1113.721495] remove   /devices/virtual/net/vethee346c56/queues/tx-3 (queues)
KERNEL[1113.721825] remove   /devices/virtual/net/vethee346c56/queues/tx-2 (queues)
KERNEL[1113.721941] remove   /devices/virtual/net/vethee346c56/queues/tx-1 (queues)
KERNEL[1113.722121] remove   /devices/virtual/net/vethee346c56/queues/rx-3 (queues)
KERNEL[1113.722196] remove   /devices/virtual/net/vethee346c56/queues/rx-2 (queues)
KERNEL[1113.722324] remove   /devices/virtual/net/vethee346c56/queues/rx-1 (queues)
KERNEL[1113.722413] remove   /devices/virtual/net/vethe5e8bca1/queues/tx-3 (queues)
KERNEL[1113.722491] remove   /devices/virtual/net/vethe5e8bca1/queues/tx-2 (queues)
KERNEL[1113.722569] remove   /devices/virtual/net/vethe5e8bca1/queues/tx-1 (queues)
KERNEL[1113.722828] remove   /devices/virtual/net/vethe5e8bca1/queues/rx-3 (queues)
KERNEL[1113.722991] remove   /devices/virtual/net/vethe5e8bca1/queues/rx-2 (queues)
KERNEL[1113.723162] remove   /devices/virtual/net/vethe5e8bca1/queues/rx-1 (queues)
UDEV  [1113.745317] add      /devices/virtual/net/vethe5e8bca1 (net)
UDEV  [1113.745365] add      /devices/virtual/net/vethee346c56 (net)
UDEV  [1113.747345] add      /devices/virtual/net/vethe5e8bca1/queues/rx-0 (queues)
UDEV  [1113.747998] add      /devices/virtual/net/vethe5e8bca1/queues/rx-2 (queues)
UDEV  [1113.748047] add      /devices/virtual/net/vethe5e8bca1/queues/rx-1 (queues)
UDEV  [1113.748848] add      /devices/virtual/net/vethe5e8bca1/queues/rx-3 (queues)
UDEV  [1113.749157] add      /devices/virtual/net/vethe5e8bca1/queues/tx-0 (queues)
UDEV  [1113.749673] add      /devices/virtual/net/vethe5e8bca1/queues/tx-1 (queues)
UDEV  [1113.750345] add      /devices/virtual/net/vethee346c56/queues/rx-3 (queues)
UDEV  [1113.750475] add      /devices/virtual/net/vethee346c56/queues/rx-1 (queues)
UDEV  [1113.750646] add      /devices/virtual/net/vethee346c56/queues/tx-0 (queues)
UDEV  [1113.750767] add      /devices/virtual/net/vethee346c56/queues/rx-0 (queues)
UDEV  [1113.750897] add      /devices/virtual/net/vethee346c56/queues/rx-2 (queues)
UDEV  [1113.751031] add      /devices/virtual/net/vethe5e8bca1/queues/tx-3 (queues)
UDEV  [1113.751227] add      /devices/virtual/net/vethee346c56/queues/tx-1 (queues)
UDEV  [1113.754159] add      /devices/virtual/net/vethee346c56/queues/tx-3 (queues)
UDEV  [1113.754604] remove   /devices/virtual/net/vethee346c56/queues/tx-1 (queues)
UDEV  [1113.754729] remove   /devices/virtual/net/vethee346c56/queues/rx-1 (queues)
UDEV  [1113.754860] remove   /devices/virtual/net/vethe5e8bca1/queues/tx-1 (queues)
UDEV  [1113.754978] remove   /devices/virtual/net/vethe5e8bca1/queues/rx-1 (queues)
UDEV  [1113.755000] remove   /devices/virtual/net/vethee346c56/queues/rx-3 (queues)
UDEV  [1113.755111] remove   /devices/virtual/net/vethee346c56/queues/rx-2 (queues)
UDEV  [1113.755230] remove   /devices/virtual/net/vethee346c56/queues/tx-3 (queues)
UDEV  [1113.755454] remove   /devices/virtual/net/vethe5e8bca1/queues/rx-3 (queues)
UDEV  [1113.755571] remove   /devices/virtual/net/vethe5e8bca1/queues/rx-2 (queues)
UDEV  [1113.755686] remove   /devices/virtual/net/vethe5e8bca1/queues/tx-3 (queues)
UDEV  [1113.755792] add      /devices/virtual/net/vethe5e8bca1/queues/tx-2 (queues)
UDEV  [1113.755919] add      /devices/virtual/net/vethee346c56/queues/tx-2 (queues)
UDEV  [1113.756740] remove   /devices/virtual/net/vethee346c56/queues/tx-2 (queues)
UDEV  [1113.756870] remove   /devices/virtual/net/vethe5e8bca1/queues/tx-2 (queues)
KERNEL[1114.024655] remove   /devices/virtual/net/vethe5e8bca1 (net)
UDEV  [1114.032461] remove   /devices/virtual/net/vethe5e8bca1 (net)
KERNEL[1115.369143] add      /class/rfkill (class)
KERNEL[1115.369443] add      /devices/virtual/misc/rfkill (misc)
KERNEL[1115.369576] add      /module/rfkill (module)
UDEV  [1115.370036] add      /class/rfkill (class)
UDEV  [1115.370130] add      /module/rfkill (module)
UDEV  [1115.370718] add      /devices/virtual/misc/rfkill (misc)
KERNEL[1115.431910] add      /class/ieee80211 (class)
KERNEL[1115.432599] add      /devices/platform/regulatory.0 (platform)
UDEV  [1115.432757] add      /class/ieee80211 (class)
UDEV  [1115.433414] add      /devices/platform/regulatory.0 (platform)
KERNEL[1115.439025] add      /module/cfg80211 (module)
UDEV  [1115.439735] add      /module/cfg80211 (module)
```

# udev

Captured while `incus stop debian-ct` is running:

```console
# udevadm monitor
monitor will print the received events for:
UDEV - the event which udev sends out after rule processing
KERNEL - the kernel uevent

KERNEL[1311.092975] add      /devices/virtual/net/eth0 (net)
KERNEL[1311.093230] move     /devices/virtual/net/physWlDc22 (net)
KERNEL[1311.093730] move     /devices/virtual/net/vethe5e8bca1 (net)
UDEV  [1311.148165] add      /devices/virtual/net/eth0 (net)
UDEV  [1311.166798] move     /devices/virtual/net/physWlDc22 (net)
UDEV  [1311.170904] move     /devices/virtual/net/vethe5e8bca1 (net)
KERNEL[1311.194217] remove   /devices/virtual/net/vethee346c56/queues/rx-0 (queues)
KERNEL[1311.194244] remove   /devices/virtual/net/vethee346c56/queues/tx-0 (queues)
KERNEL[1311.194268] remove   /devices/virtual/net/vethee346c56 (net)
KERNEL[1311.194295] remove   /devices/virtual/net/vethe5e8bca1/queues/rx-0 (queues)
KERNEL[1311.194305] remove   /devices/virtual/net/vethe5e8bca1/queues/tx-0 (queues)
KERNEL[1311.194344] remove   /devices/virtual/net/vethe5e8bca1 (net)
UDEV  [1311.195050] remove   /devices/virtual/net/vethee346c56/queues/rx-0 (queues)
UDEV  [1311.195142] remove   /devices/virtual/net/vethee346c56/queues/tx-0 (queues)
UDEV  [1311.195349] remove   /devices/virtual/net/vethe5e8bca1/queues/rx-0 (queues)
UDEV  [1311.195945] remove   /devices/virtual/net/vethe5e8bca1/queues/tx-0 (queues)
UDEV  [1311.198020] remove   /devices/virtual/net/vethee346c56 (net)
UDEV  [1311.200124] remove   /devices/virtual/net/vethe5e8bca1 (net)
```

**Comment by DanielG** [comment]
Kernel bug report: https://lists.sr.ht/~dxld/outbox/%3C20231010121003.x3yi6fihecewjy4e%40House.clients.dxld.at%3E

**Comment by DanielG** [comment]
@stgraber could you reopen this issue so it doesn't get lost?

Upstream seems to agree this is an issue and even more bugs fell out when taking off the covers. `ip monitor link` also gets bogus info from the kernel here.

I would recommend working around this issue by renaming the device inside the netns before moving it back to the main-netns. Is there likely to be a udev running inside the containers? In that case perhaps care should be taken to terminate all processes before doing the move so as to not confuse it.

**Comment by stgraber** [comment]
@DanielG can you file a bug at https://github.com/lxc/lxc/issues as that's where the logic moving the interface to the host resides

**Comment by DanielG** [comment]
@stgraber I don't want to fragment the bug history, so I'm not going to.

FYI Jakub has posted a fix to netdev and it has landed in net-next. https://patchwork.kernel.org/project/netdevbpf/patch/20231120184140.578375-1-kuba@kernel.org/

**Comment by ahmgithubahm** [comment]
Just wondered if there's a workaround for this on Debian 12 with Incus LTS 6.0.1 from the Bookworm backports repo ?  I am seeing the same with a new VPS (from BuyVM.net).

**Comment by stgraber** [comment]
Make sure your kernel is up to date. The kernel fix should have been picked up by all distros by now.

**Comment by ahmgithubahm** [comment]
Thanks Stéphane.  I'm on 6.1.99, but Debian 12 backports has 6.9.7 - sounds recent enough? :-)

**Comment by atomlab** [comment]
Debian 12 with kernel 6.1.0-23.

The same problem with removing IP addresses from the eth0 interface when I stop the container.
```
incus stop —force container_name
```

Indeed, the problem was solved by updating the kernel to version 6.9.7 from the backports repository.

```
apt install linux-image-amd64/bookworm-backports
```

```
# uname -r
6.9.7+bpo-amd64
```

Thanks!
